diff --git a/src/code/z_demo.c b/src/code/z_demo.c
index 358edd32e..c39e1f8e9 100644
--- a/src/code/z_demo.c
+++ b/src/code/z_demo.c
@@ -188,6 +188,7 @@ void Cutscene_DrawDebugInfo(PlayState* play, Gfx** dlist, CutsceneContext* csCtx
 void Cutscene_InitContext(PlayState* play, CutsceneContext* csCtx) {
     csCtx->state = CS_STATE_IDLE;
     csCtx->timer = 0.0f;
+    csCtx->skippable = false;
 }
 
 void Cutscene_StartManual(PlayState* play, CutsceneContext* csCtx) {
@@ -590,6 +591,10 @@ void CutsceneCmd_Destination(PlayState* play, CutsceneContext* csCtx, CsCmdDesti
     Player* player = GET_PLAYER(play);
     s32 titleDemoSkipped = false;
 
+    if (!csCtx->skippable && ((csCtx->curFrame > 30) && (gSaveContext.fileNum != 0xFEDC) && (gSaveContext.fileNum != 0xFF))) {
+        csCtx->skippable = true;
+    }
+
     if ((gSaveContext.gameMode != GAMEMODE_NORMAL) && (gSaveContext.gameMode != GAMEMODE_END_CREDITS) &&
         (play->sceneId != SCENE_HYRULE_FIELD) && (csCtx->curFrame > 20) &&
         (CHECK_BTN_ALL(play->state.input[0].press.button, BTN_A) ||
@@ -601,8 +606,7 @@ void CutsceneCmd_Destination(PlayState* play, CutsceneContext* csCtx, CsCmdDesti
     }
 
     if ((csCtx->curFrame == cmd->startFrame) || titleDemoSkipped ||
-        (DEBUG_FEATURES && (csCtx->curFrame > 20) && CHECK_BTN_ALL(play->state.input[0].press.button, BTN_START) &&
-         (gSaveContext.fileNum != 0xFEDC))) {
+        (csCtx->skippable && CHECK_BTN_ALL(play->state.input[0].press.button, BTN_START))) {
         csCtx->state = CS_STATE_RUN_UNSTOPPABLE;
         Audio_SetCutsceneFlag(0);
         gSaveContext.cutsceneTransitionControl = 1;
@@ -902,6 +906,7 @@ void CutsceneCmd_Destination(PlayState* play, CutsceneContext* csCtx, CsCmdDesti
                 break;
 
             case CS_DEST_LAKE_HYLIA_FROM_LAKE_RESTORED:
+                SET_EVENTCHKINF(EVENTCHKINF_RESTORED_LAKE_HYLIA);
                 play->nextEntranceIndex = ENTR_LAKE_HYLIA_5;
                 play->transitionTrigger = TRANS_TRIGGER_START;
                 play->transitionType = TRANS_TYPE_FADE_BLACK;
@@ -915,6 +920,7 @@ void CutsceneCmd_Destination(PlayState* play, CutsceneContext* csCtx, CsCmdDesti
                 break;
 
             case CS_DEST_WINDMILL_FROM_WELL_DRAINED:
+                SET_EVENTCHKINF(EVENTCHKINF_DRAINED_WELL);
                 play->nextEntranceIndex = ENTR_WINDMILL_AND_DAMPES_GRAVE_2;
                 play->transitionTrigger = TRANS_TRIGGER_START;
                 play->transitionType = TRANS_TYPE_FADE_BLACK_FAST;
@@ -945,6 +951,7 @@ void CutsceneCmd_Destination(PlayState* play, CutsceneContext* csCtx, CsCmdDesti
                 break;
 
             case CS_DEST_DESERT_COLOSSUS_FROM_REQUIEM:
+                Item_Give(play, ITEM_SONG_REQUIEM);
                 play->nextEntranceIndex = ENTR_DESERT_COLOSSUS_4;
                 play->transitionTrigger = TRANS_TRIGGER_START;
                 play->transitionType = TRANS_TYPE_SANDSTORM_END;
@@ -2405,15 +2412,20 @@ void Cutscene_HandleEntranceTriggers(PlayState* play) {
             requiredAge = gSaveContext.save.linkAge;
         }
 
-        if ((gSaveContext.save.entranceIndex == entranceCutscene->entrance) &&
-            (!Flags_GetEventChkInf(entranceCutscene->flag) || (entranceCutscene->flag == EVENTCHKINF_EPONA_OBTAINED)) &&
+        if ((gSaveContext.save.entranceIndex == entranceCutscene->entrance)) {
+            if (entranceCutscene->entrance == ENTR_HYRULE_FIELD_16) {
+                Item_Give(play, ITEM_SONG_TIME);
+            }
+
+            if ((!Flags_GetEventChkInf(entranceCutscene->flag) || (entranceCutscene->flag == EVENTCHKINF_EPONA_OBTAINED)) &&
             (gSaveContext.save.cutsceneIndex < CS_INDEX_0) && ((u8)gSaveContext.save.linkAge == requiredAge) &&
             (gSaveContext.respawnFlag <= 0)) {
-            Flags_SetEventChkInf(entranceCutscene->flag);
-            Cutscene_SetScript(play, entranceCutscene->script);
-            gSaveContext.cutsceneTrigger = 2;
-            gSaveContext.showTitleCard = false;
-            break;
+                Flags_SetEventChkInf(entranceCutscene->flag);
+                Cutscene_SetScript(play, entranceCutscene->script);
+                gSaveContext.cutsceneTrigger = 2;
+                gSaveContext.showTitleCard = false;
+                break;
+            }
         }
     }
 }
@@ -2441,6 +2453,7 @@ void Cutscene_HandleConditionalTriggers(PlayState* play) {
                    LINK_IS_ADULT && !Flags_GetEventChkInf(EVENTCHKINF_C4) &&
                    (gEntranceTable[((void)0, gSaveContext.save.entranceIndex)].sceneId == SCENE_TEMPLE_OF_TIME)) {
             Flags_SetEventChkInf(EVENTCHKINF_C4);
+            Item_Give(play, ITEM_ARROW_LIGHT);
             gSaveContext.save.entranceIndex = ENTR_TEMPLE_OF_TIME_0;
             gSaveContext.save.cutsceneIndex = CS_INDEX_8;
         } else if (!Flags_GetEventChkInf(EVENTCHKINF_C7) &&
