diff --git a/src/gcc_fix/missing_gcc_functions.c b/src/gcc_fix/missing_gcc_functions.c
index 615dd2354..22f83d854 100644
--- a/src/gcc_fix/missing_gcc_functions.c
+++ b/src/gcc_fix/missing_gcc_functions.c
@@ -26,6 +26,12 @@ int memcmp(const void* s1, const void* s2, size_t n) {
     return 0;
 }
 
+char* strcpy (char *dst, const char *src)
+{
+    const size_t len = strlen (src);
+    return (char *) memcpy (dst, src, len + 1) + len;
+}
+
 // Conversions involving 64-bit integer types required by the O32 MIPS ABI.
 
 // f32 -> u64, negative values become 0
