diff --git a/src/overlays/actors/ovl_En_Box/z_en_box.c b/src/overlays/actors/ovl_En_Box/z_en_box.c
index c4ef15919..eb1e4a94a 100644
--- a/src/overlays/actors/ovl_En_Box/z_en_box.c
+++ b/src/overlays/actors/ovl_En_Box/z_en_box.c
@@ -7,6 +7,7 @@
 #include "ichain.h"
 #include "one_point_cutscene.h"
 #include "printf.h"
+#include "randomizer_play.h"
 #include "sequence.h"
 #include "sfx.h"
 #include "sys_math3d.h"
@@ -109,6 +110,14 @@ void EnBox_ClipToGround(EnBox* this, PlayState* play) {
     }
 }

+GetItemID EnBox_GetContentsGetItemID(EnBox* this, PlayState* play) {
+    unsigned flag = PARAMS_GET_U(this->dyna.actor.params, 0, 5);
+    if (gSaveContext.save.randomizer.shuffleChests && !(play->sceneId == SCENE_TREASURE_BOX_SHOP && flag < 0x0A)) {
+        return gSaveContext.save.randomizer.chestContents[Randomizer_GetRandomizedChestContents(play->sceneId, flag)];
+    }
+    return PARAMS_GET_U(this->dyna.actor.params, 5, 7);
+}
+
 void EnBox_Init(Actor* thisx, PlayState* play2) {
     PlayState* play = play2;
     EnBox* this = (EnBox*)thisx;
@@ -195,6 +204,13 @@ void EnBox_Init(Actor* thisx, PlayState* play2) {
                    this->morphTable, 5);
     Animation_Change(&this->skelanime, anim, 1.5f, animFrameStart, endFrame, ANIMMODE_ONCE, 0.0f);

+    if (EnBox_GetContentsGetItemID(this, play) == GI_ICE_TRAP) {
+        // Spawn the object for the ice trap wind effect if the scene has not loaded it
+        if (Object_GetSlot(&play->objectCtx, OBJECT_FZ) == -1) {
+            Object_SpawnTemporary(&play->objectCtx, OBJECT_FZ);
+        }
+    }
+
     switch (this->type) {
         case ENBOX_TYPE_SMALL:
         case ENBOX_TYPE_6:
@@ -446,7 +462,7 @@ void EnBox_WaitOpen(EnBox* this, PlayState* play) {
         Actor_WorldToActorCoords(&this->dyna.actor, &sp4C, &player->actor.world.pos);
         if (sp4C.z > -50.0f && sp4C.z < 0.0f && fabsf(sp4C.y) < 10.0f && fabsf(sp4C.x) < 20.0f &&
             Player_IsFacingActor(&this->dyna.actor, 0x3000, play)) {
-            Actor_OfferGetItemNearby(&this->dyna.actor, play, -PARAMS_GET_U(this->dyna.actor.params, 5, 7));
+            Actor_OfferGetItemNearby(&this->dyna.actor, play, -EnBox_GetContentsGetItemID(this, play));
         }
         if (Flags_GetTreasure(play, PARAMS_GET_U(this->dyna.actor.params, 0, 5))) {
             EnBox_SetupAction(this, EnBox_Open);
@@ -559,7 +575,7 @@ void EnBox_Update(Actor* thisx, PlayState* play) {
             Actor_SetFocus(&this->dyna.actor, 40.0f);
     }

-    if (PARAMS_GET_U(this->dyna.actor.params, 5, 7) == GI_ICE_TRAP && this->actionFunc == EnBox_Open &&
+    if (EnBox_GetContentsGetItemID(this, play) == GI_ICE_TRAP && this->actionFunc == EnBox_Open &&
         this->skelanime.curFrame > 45 && this->iceSmokeTimer < 100) {
         EnBox_SpawnIceSmoke(this, play);
     }