diff --git a/src/overlays/actors/ovl_En_Mag/z_en_mag.c b/src/overlays/actors/ovl_En_Mag/z_en_mag.c
index 5c1e68f77..be0a87742 100644
--- a/src/overlays/actors/ovl_En_Mag/z_en_mag.c
+++ b/src/overlays/actors/ovl_En_Mag/z_en_mag.c
@@ -74,7 +74,7 @@ void EnMag_ResetSram(void) {
     SsSram_ReadWrite(OS_K1_TO_PHYSICAL(0xA8007000), buffer, 0x800, 1);
     SsSram_ReadWrite(OS_K1_TO_PHYSICAL(0xA8007800), buffer, 0x800, 1);

-    gSaveContext.soundSetting = gSaveContext.zTargetSetting = 0; // SOUND_SETTING_STEREO/Z_TARGET_SETTING_SWITCH
+    gSaveContext.quicktextSetting = gSaveContext.soundSetting = gSaveContext.zTargetSetting = 0; // QUICKTEXT_SETTING_OFF/SOUND_SETTING_STEREO/Z_TARGET_SETTING_SWITCH
     Audio_SetSoundOutputMode(gSaveContext.soundSetting);
 }
 #endif
@@ -571,6 +571,18 @@ void EnMag_DrawInner(Actor* thisx, PlayState* play, Gfx** gfxP) {
         FILENAME_UPPERCASE('R'),
         FILENAME_UPPERCASE('T'),
     };
+    static u8 randomizerFontIndices[] = {
+        FILENAME_UPPERCASE('R'),
+        FILENAME_UPPERCASE('A'),
+        FILENAME_UPPERCASE('N'),
+        FILENAME_UPPERCASE('D'),
+        FILENAME_UPPERCASE('O'),
+        FILENAME_UPPERCASE('M'),
+        FILENAME_UPPERCASE('I'),
+        FILENAME_UPPERCASE('Z'),
+        FILENAME_UPPERCASE('E'),
+        FILENAME_UPPERCASE('R'),
+    };
     static void* effectMaskTextures[] = {
         gTitleEffectMask00Tex, gTitleEffectMask01Tex, gTitleEffectMask02Tex,
         gTitleEffectMask10Tex, gTitleEffectMask11Tex, gTitleEffectMask12Tex,
@@ -657,6 +669,49 @@ void EnMag_DrawInner(Actor* thisx, PlayState* play, Gfx** gfxP) {
         EnMag_DrawTextureI8(&gfx, gTitleTheLegendOfTextTex, 72, 8, 153 + LOGO_X_SHIFT, 72, 72, 8, 1024, 1024);
         EnMag_DrawTextureI8(&gfx, gTitleOcarinaOfTimeTMTextTex, 96, 8, 151 + LOGO_X_SHIFT, 126, 96, 8, 1024, 1024);

+        // Draw "RANDOMIZER" Text
+        // Shadow
+        gDPPipeSync(gfx++);
+        gDPSetPrimColor(gfx++, 0, 0, 0, 0, 0, this->mainAlpha);
+
+        rectLeft = YREG(7) + 40 + 1;
+        for (i = 0; i < ARRAY_COUNT(randomizerFontIndices); i++) {
+            EnMag_DrawCharTexture(&gfx, font->fontBuf + randomizerFontIndices[i] * FONT_CHAR_TEX_SIZE, rectLeft, YREG(10) + 131 + 1);
+            rectLeft += YREG(8);
+            if (i == 0) {
+                rectLeft -= 1;
+            } else if (i == 4) {
+                rectLeft += 1;
+            } else if (i == 5) {
+                rectLeft += 2;
+            } else if (i == 6) {
+                rectLeft -= 4;
+            } else if (i == 8) {
+                rectLeft -= 1;
+            }
+        }
+
+        // Actual
+        gDPPipeSync(gfx++);
+        gDPSetPrimColor(gfx++, 0, 0, 100, 150, 255, this->mainAlpha);
+
+        rectLeft = YREG(7) + 40;
+        for (i = 0; i < ARRAY_COUNT(randomizerFontIndices); i++) {
+            EnMag_DrawCharTexture(&gfx, font->fontBuf + randomizerFontIndices[i] * FONT_CHAR_TEX_SIZE, rectLeft, YREG(10) + 131);
+            rectLeft += YREG(8);
+            if (i == 0) {
+                rectLeft -= 1;
+            } else if (i == 4) {
+                rectLeft += 1;
+            } else if (i == 5) {
+                rectLeft += 2;
+            } else if (i == 6) {
+                rectLeft -= 4;
+            } else if (i == 8) {
+                rectLeft -= 1;
+            }
+        }
+
 #if OOT_MQ
         gDPPipeSync(gfx++);
         gDPSetPrimColor(gfx++, 0, 0, 255, 255, 255, (s16)this->subAlpha);