diff --git a/src/overlays/gamestates/ovl_file_choose/z_file_copy_erase.c b/src/overlays/gamestates/ovl_file_choose/z_file_copy_erase.c
index 03fb5d0f1..78b98e7e6 100644
--- a/src/overlays/gamestates/ovl_file_choose/z_file_copy_erase.c
+++ b/src/overlays/gamestates/ovl_file_choose/z_file_copy_erase.c
@@ -13,12 +13,11 @@
 // compiler errors (see docs/compilers.md)

 // when choosing a file to copy or erase, the 6 main menu buttons are placed at these offsets
-static s16 sChooseFileYOffsets[] = { -48, -48, -48, -24, -24, 0 };
+static s16 sChooseFileYOffsets[] = { -48, -48, -24, -24, 0 };

 static s16 D_8081248C[3][3] = {
-    { 0, -48, -48 },
-    { -64, 16, -48 },
-    { -64, -64, 32 },
+    { 0, -48, },
+    { -64, 16 },
 };

 static s16 sEraseDelayTimer = 15;
@@ -36,7 +35,7 @@ void FileSelect_SetupCopySource(GameState* thisx) {
 #endif

 #if !PLATFORM_IQUE
-    for (i = 0; i < 5; i++) {
+    for (i = 0; i < 4; i++) {
         yStep = (ABS(this->buttonYOffsets[i] - sChooseFileYOffsets[i])) / this->actionTimer;

         if (this->buttonYOffsets[i] >= sChooseFileYOffsets[i]) {
@@ -157,7 +156,7 @@ void FileSelect_SetupCopyDest1(GameState* thisx) {
 #endif

 #if !PLATFORM_IQUE
-    for (i = 0; i < 3; i++) {
+    for (i = 0; i < 2; i++) {
         yStep = ABS(this->buttonYOffsets[i] - D_8081248C[this->buttonIndex][i]) / this->actionTimer;

         if (this->buttonYOffsets[i] <= D_8081248C[this->buttonIndex][i]) {
@@ -383,7 +382,7 @@ void FileSelect_ExitToCopySource2(GameState* thisx) {
 #endif

 #if !PLATFORM_IQUE
-    for (i = 0; i < 3; i++) {
+    for (i = 0; i < 2; i++) {
         yStep = ABS(this->buttonYOffsets[i] - sChooseFileYOffsets[i]) / this->actionTimer;

         if (this->buttonYOffsets[i] >= sChooseFileYOffsets[i]) {
@@ -438,7 +437,7 @@ void FileSelect_ExitToCopySource2(GameState* thisx) {
  * Update function for `CM_SETUP_COPY_CONFIRM_1`
  */
 void FileSelect_SetupCopyConfirm1(GameState* thisx) {
-    static s16 D_808124A4[] = { -56, -40, -24, 0 };
+    static s16 D_808124A4[] = { -56, -40, 0 };
     FileSelectState* this = (FileSelectState*)thisx;
     SramContext* sramCtx = &this->sramCtx;
     s16 i;
@@ -457,7 +456,7 @@ void FileSelect_SetupCopyConfirm1(GameState* thisx) {
 #endif

 #if !PLATFORM_IQUE
-    for (i = 0; i < 3; i++) {
+    for (i = 0; i < 2; i++) {
         if ((i != this->copyDestFileIndex) && (i != this->selectedFileIndex)) {
             this->fileButtonAlpha[i] -= 25;

@@ -599,7 +598,7 @@ void FileSelect_ReturnToCopyDest(GameState* thisx) {
 #endif

 #if !PLATFORM_IQUE
-    for (i = 0; i < 3; i++) {
+    for (i = 0; i < 2; i++) {
         if ((i != this->copyDestFileIndex) && (i != this->selectedFileIndex)) {
             this->fileButtonAlpha[i] += 25;

@@ -854,7 +853,7 @@ void FileSelect_CopyAnim5(GameState* thisx) {
 #endif

 #if !PLATFORM_IQUE
-    for (i = 0; i < 3; i++) {
+    for (i = 0; i < 2; i++) {
         if (i != this->buttonIndex) {
             this->fileButtonAlpha[i] += 25;

@@ -902,7 +901,7 @@ void FileSelect_CopyAnim5(GameState* thisx) {

     if (this->actionTimer == 0) {
 #if !PLATFORM_IQUE
-        for (i = 0; i < 3; i++) {
+        for (i = 0; i < 2; i++) {
             this->fileButtonAlpha[i] = 200;
             this->nameBoxAlpha[i] = this->nameAlpha[i] = this->connectorAlpha[i] = 0;

@@ -1215,7 +1214,7 @@ void FileSelect_SetupEraseConfirm1(GameState* thisx) {
 #endif

 #if !PLATFORM_IQUE
-    for (i = 0; i < 3; i++) {
+    for (i = 0; i < 2; i++) {
         if (i != this->buttonIndex) {
             this->fileButtonAlpha[i] -= 25;

@@ -1282,7 +1281,7 @@ void FileSelect_SetupEraseConfirm1(GameState* thisx) {

     if (this->actionTimer == 0) {
 #if !PLATFORM_IQUE
-        for (i = 0; i < 3; i++) {
+        for (i = 0; i < 2; i++) {
             if (i != this->buttonIndex) {
                 this->fileButtonAlpha[i] = 0;

@@ -1481,7 +1480,7 @@ void FileSelect_ExitToEraseSelect2(GameState* thisx) {
 #endif

 #if !PLATFORM_IQUE
-    for (i = 0; i < 3; i++) {
+    for (i = 0; i < 2; i++) {
         if (i != this->buttonIndex) {
             this->fileButtonAlpha[i] += 25;

@@ -1683,7 +1682,7 @@ void FileSelect_EraseAnim3(GameState* thisx) {
 #endif

 #if !PLATFORM_IQUE
-    for (i = 0; i < 3; i++) {
+    for (i = 0; i < 2; i++) {
         this->fileButtonAlpha[i] += 25;

         if (SLOT_OCCUPIED(sramCtx, i)) {