diff --git a/src/boot/idle.c b/src/boot/idle.c
index 648ce48cf..230f5a5bc 100644
--- a/src/boot/idle.c
+++ b/src/boot/idle.c
@@ -92,7 +92,7 @@ void Idle_ThreadEntry(void* arg) {
     gViConfigYScale = 1.0f;
 #endif
 
-#if DEBUG_FEATURES
+#if 0
     // Allow both 60 Hz and 50 Hz
     switch (osTvType) {
         case OS_TV_NTSC:
@@ -113,7 +113,7 @@ void Idle_ThreadEntry(void* arg) {
 #endif
             break;
     }
-#elif !OOT_PAL_N64
+#elif OOT_REGION != REGION_EU
     // 60 Hz only (GameCube and NTSC N64)
     switch (osTvType) {
         case OS_TV_PAL:
