diff --git a/include/sram.h b/include/sram.h
index f683be8a0..94dfc73cb 100644
--- a/include/sram.h
+++ b/include/sram.h
@@ -11,12 +11,13 @@ typedef struct SramContext {
 } SramContext; // size = 0x4
 
 #define SRAM_SIZE 0x8000
-#define SRAM_HEADER_SIZE 0x10
+#define SRAM_HEADER_SIZE 0x11
 
 typedef enum SramHeaderField {
     /* 0x00 */ SRAM_HEADER_SOUND,
     /* 0x01 */ SRAM_HEADER_Z_TARGET,
     /* 0x02 */ SRAM_HEADER_LANGUAGE,
+               SRAM_HEADER_QUICKTEXT,
     /* 0x03 */ SRAM_HEADER_MAGIC // must be the value in `sSramDefaultHeader` for save to be considered valid
 } SramHeaderField;
 
