diff --git a/include/z64cutscene.h b/include/z64cutscene.h
index f8aed368d..15086e187 100644
--- a/include/z64cutscene.h
+++ b/include/z64cutscene.h
@@ -499,7 +499,7 @@ typedef struct CutsceneCameraMove {
 } CutsceneCameraMove; // size = 0xC

 typedef struct CutsceneContext {
-    /* 0x00 */ char  unk_00[0x4];
+    /* 0x00 */ char  unk_00[0x3];
     /* 0x04 */ void* script;
     /* 0x08 */ u8 state;
     /* 0x0C */ f32 timer;
@@ -513,6 +513,7 @@ typedef struct CutsceneContext {
     /* 0x20 */ CutsceneCameraPoint* camEyePoints;
     /* 0x24 */ CsCmdActorCue* playerCue;
     /* 0x28 */ CsCmdActorCue* actorCues[10]; // "npcdemopnt"
+    bool skippable;
 } CutsceneContext; // size = 0x50

 extern u8 gUseCutsceneCam;