diff --git a/src/libc64/qrand.c b/src/libc64/qrand.c
index 69bfa11d7..6655e46c1 100644
--- a/src/libc64/qrand.c
+++ b/src/libc64/qrand.c
@@ -79,6 +79,21 @@ u32 Rand_Next(void) {
 #endif
 }

+u32 Rand_Next_Mod(u32 mod) {
+    if (mod == 0) {
+        return 0;
+    }
+
+    u8 count = 0;
+    u32 shift = mod;
+    while ((shift & 0x80000000) == 0) {
+        count++;
+        shift <<= 1;
+    }
+
+    return (Rand_Next() >> count) % mod;
+}
+
 /**
  * Seeds the pseudo-random number generator by providing a starting value.
  *
@@ -156,6 +171,21 @@ u32 Rand_Next_Variable(u32* rndNum) {
     return *rndNum = (*rndNum) * RAND_MULTIPLIER + RAND_INCREMENT;
 }

+u32 Rand_Next_Mod_Variable(u32* rndNum, u32 mod) {
+    if (mod == 0) {
+        return 0;
+    }
+
+    u8 count = 0;
+    u32 shift = mod;
+    while ((shift & 0x80000000) == 0) {
+        count++;
+        shift <<= 1;
+    }
+
+    return *rndNum = (*rndNum >> count) % mod;
+}
+
 /**
  * Generates the next pseudo-random floating-point number between 0.0f and 1.0f from the provided rndNum.
  *